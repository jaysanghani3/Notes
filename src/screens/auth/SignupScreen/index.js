import React, {useState} from 'react';
import {Text, View, TouchableOpacity, TextInput, Image} from 'react-native';
import firestore from '@react-native-firebase/firestore';
import auth from '@react-native-firebase/auth';
import styles from './styles';
import globalStyles from '../../../utils/globalStyles';
import Toast from 'react-native-simple-toast';
import Loader from '../../../components/Loader';
import Icons from '../../../assets';
import Strings from '../../../translation/Strings';
import {
  isEmailValid,
  isNameValid,
  isMobileNumberValid,
  isConfirmPasswordValid,
  isPasswordStrong,
  isPasswordMatched,
  isNameContainsNumber,
} from '../../../utils/validations';

const SignupScreen = ({navigation}) => {
  const fields = [
    {
      title: 'fullName',
      value: 'Full Name',
      type: 'text',
      errorMsg: 'Please enter your Name',
    },
    {
      title: 'mobileNumber',
      value: 'Mobile Number',
      type: 'tel',
      errorMsg: 'Please enter Mobile Number',
    },
    {
      title: 'email',
      value: 'Email',
      type: 'email',
      errorMsg: 'Please enter Email',
    },

    {
      title: 'password',
      value: 'Password',
      type: 'text',
      errorMsg: 'Please Fill Password',
    },
    {
      title: 'confirmPassword',
      value: 'Confirm Password',
      type: 'text',
      errorMsg: 'Please fill Password or Password not matched ',
    },
  ];

  const [userDetails, setUserDetails] = useState({
    fullName: '',
    mobileNumber: '',
    email: '',
    password: '',
    confirmPassword: '',
  });
  const [error, setError] = useState({});
  const [isLoading, setIsLoading] = useState(false);
  const [isPasswordVisible, setIsPasswordVisible] = useState(false);
  const [isConfirmPasswordVisible, setIsConfirmPasswordVisible] =
    useState(false);

  const handlePress = async () => {
    try {
      if (!userDetails.fullName) {
        setError({fullName: Strings.required});
        return;
      }
      if (isNameContainsNumber(userDetails.fullName)) {
        setError({fullName: Strings.charName});
        return;
      }
      if (!isNameValid(userDetails.fullName)) {
        setError({fullName: Strings.invalidName});
        return;
      }

      if (!userDetails.mobileNumber) {
        setError({mobileNumber: Strings.required});
        return;
      }
      if (!isMobileNumberValid(userDetails.mobileNumber)) {
        setError({mobileNumber: Strings.invalidPhone});
        return;
      }

      if (!userDetails.email) {
        setError({email: Strings.required});
        return;
      }
      if (!isEmailValid(userDetails.email)) {
        setError({email: Strings.invalidEmail});
        return;
      }
      if (!userDetails.password) {
        setError({password: Strings.required});
        return;
      }

      if (!userDetails.confirmPassword) {
        setError({confirmPassword: Strings.required});
        return;
      }
      if (
        !isPasswordMatched(userDetails.password, userDetails.confirmPassword)
      ) {
        setError({confirmPassword: Strings.invalidPasswordConfirmation});
        return;
      }
      if (!isPasswordStrong(userDetails.password)) {
        setError({password: Strings.invalidPassword});
        return;
      }
      if (
        !isConfirmPasswordValid(
          userDetails.password,
          userDetails.confirmPassword,
        )
      ) {
        setError({confirmPassword: Strings.invalidPasswordConfirmation});
        return;
      }
      setError({});

      setIsLoading(true);
      const user = await auth().createUserWithEmailAndPassword(
        userDetails.email,
        userDetails.password,
      );
      await user.user.sendEmailVerification();
      await firestore().collection('users').doc(user.user.uid).set({
        fullName: userDetails.fullName,
        mobileNumber: userDetails.mobileNumber,
        email: userDetails.email,
      });
      setIsLoading(false);
      Toast.show(Strings.linkSend);
      navigation.navigate('LoginScreen');
    } catch (e) {
      const err = e.code.split('/')[1].split('-').join(' ');
      Toast.show(err);
      setIsLoading(false);
    }
  };

  return (
    <View style={globalStyles.container}>
      <Loader state={isLoading} />

      <View>
        <Text style={styles.welcomeHeader}>{Strings.createAccount}</Text>
        <Text>{Strings.signupToGetStarted}</Text>
      </View>

      <View style={styles.formBox}>
        {fields.map((field, index) => (
          <View key={index} style={styles.fieldArea}>
            <View style={styles.textArea}>
              <TextInput
                inputMode={field.type}
                key={index}
                defaultValue={userDetails[field.title]}
                onChangeText={newValue => {
                  setUserDetails({...userDetails, [field.title]: newValue});
                }}
                style={globalStyles.flex1}
                placeholder={field.value}
                secureTextEntry={
                  field.title === 'password'
                    ? !isPasswordVisible
                    : field.title === 'confirmPassword' &&
                      !isConfirmPasswordVisible
                }
              />
              {field.title === 'password' && (
                <TouchableOpacity
                  style={styles.eyeContainer}
                  onPress={() => setIsPasswordVisible(!isPasswordVisible)}>
                  <Image
                    source={!isPasswordVisible ? Icons.Show : Icons.Hide}
                    style={styles.eye}
                  />
                </TouchableOpacity>
              )}
              {field.title === 'confirmPassword' && (
                <TouchableOpacity
                  style={styles.eyeContainer}
                  onPress={() =>
                    setIsConfirmPasswordVisible(!isConfirmPasswordVisible)
                  }>
                  <Image
                    source={!isConfirmPasswordVisible ? Icons.Show : Icons.Hide}
                    style={styles.eye}
                  />
                </TouchableOpacity>
              )}
            </View>
            {error[field.title] !== '' && (
              <Text style={globalStyles.error}>{error[field.title]}</Text>
            )}
          </View>
        ))}
      </View>

      <View>
        <TouchableOpacity onPress={handlePress} style={globalStyles.btn}>
          <Text style={globalStyles.btnText}>{Strings.signupBtn}</Text>
        </TouchableOpacity>
      </View>

      <View style={globalStyles.topBorder}>
        <Text
          style={globalStyles.textCenter}
          onPress={() => navigation.navigate('LoginScreen')}>
          {Strings.alreadyHaveAccount}
          <Text style={globalStyles.violetText}> {Strings.loginBtn}</Text>
        </Text>
      </View>
    </View>
  );
};

export default SignupScreen;
