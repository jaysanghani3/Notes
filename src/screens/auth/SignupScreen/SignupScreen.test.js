import SignupScreen from './index';
import React from 'react';
import renderer from 'react-test-renderer';
import {test, expect} from '@jest/globals';

// jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
test('renders correctly', () => {
  const tree = renderer.create(<SignupScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
