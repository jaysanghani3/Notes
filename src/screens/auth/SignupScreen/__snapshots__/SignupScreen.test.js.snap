// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "justifyContent": "center",
      "padding": 25,
    }
  }
>
  <View>
    <Text
      style={
        {
          "color": "#000000",
          "fontSize": 30,
          "fontWeight": "bold",
        }
      }
    >
      Create an account
    </Text>
    <Text>
      Signup to get started.
    </Text>
  </View>
  <View
    style={
      {
        "marginVertical": 40,
      }
    }
  >
    <View
      style={
        {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#B6B6B6",
            "borderRadius": 8,
            "borderWidth": 1,
            "display": "block",
            "flexDirection": "row",
            "paddingStart": 10,
          }
        }
      >
        <TextInput
          defaultValue=""
          inputMode="text"
          onChangeText={[Function]}
          placeholder="Full Name"
          secureTextEntry={false}
          style={
            {
              "flex": 1,
            }
          }
        />
      </View>
      <Text
        style={
          {
            "color": "#D32F2F",
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#B6B6B6",
            "borderRadius": 8,
            "borderWidth": 1,
            "display": "block",
            "flexDirection": "row",
            "paddingStart": 10,
          }
        }
      >
        <TextInput
          defaultValue=""
          inputMode="tel"
          onChangeText={[Function]}
          placeholder="Mobile Number"
          secureTextEntry={false}
          style={
            {
              "flex": 1,
            }
          }
        />
      </View>
      <Text
        style={
          {
            "color": "#D32F2F",
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#B6B6B6",
            "borderRadius": 8,
            "borderWidth": 1,
            "display": "block",
            "flexDirection": "row",
            "paddingStart": 10,
          }
        }
      >
        <TextInput
          defaultValue=""
          inputMode="email"
          onChangeText={[Function]}
          placeholder="Email"
          secureTextEntry={false}
          style={
            {
              "flex": 1,
            }
          }
        />
      </View>
      <Text
        style={
          {
            "color": "#D32F2F",
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#B6B6B6",
            "borderRadius": 8,
            "borderWidth": 1,
            "display": "block",
            "flexDirection": "row",
            "paddingStart": 10,
          }
        }
      >
        <TextInput
          defaultValue=""
          inputMode="text"
          onChangeText={[Function]}
          placeholder="Password"
          secureTextEntry={true}
          style={
            {
              "flex": 1,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../src/assets/show.png",
              }
            }
            style={
              {
                "height": 25,
                "position": "absolute",
                "right": 15,
                "top": 11,
                "width": 25,
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#D32F2F",
          }
        }
      />
    </View>
    <View
      style={
        {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#B6B6B6",
            "borderRadius": 8,
            "borderWidth": 1,
            "display": "block",
            "flexDirection": "row",
            "paddingStart": 10,
          }
        }
      >
        <TextInput
          defaultValue=""
          inputMode="text"
          onChangeText={[Function]}
          placeholder="Confirm Password"
          secureTextEntry={true}
          style={
            {
              "flex": 1,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../src/assets/show.png",
              }
            }
            style={
              {
                "height": 25,
                "position": "absolute",
                "right": 15,
                "top": 11,
                "width": 25,
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#D32F2F",
          }
        }
      />
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#6574FF",
          "borderRadius": 8,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Sign Up
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "borderTopColor": "#B6B6B6",
        "borderTopWidth": 1,
        "marginTop": 50,
        "padding": 10,
      }
    }
  >
    <Text
      onPress={[Function]}
      style={
        {
          "textAlign": "center",
        }
      }
    >
      Already have an account? 
      <Text
        style={
          {
            "color": "#6574FF",
          }
        }
      >
         
        Login
      </Text>
    </Text>
  </View>
</View>
`;
